From 3d8eff0a21cc02569807f95dd5697724034a3e0a Mon Sep 17 00:00:00 2001
From: Jung Zhao <jung.zhao@rock-chips.com>
Date: Sat, 24 Feb 2018 17:24:05 +0800
Subject: [PATCH] video: rockchip: vpu: fix wrong freq assigned

since only rk3288 can run on 600MHz, and sometimes other socs
will step into the 600MHz branch when its 3th register value can
meet the condition, it will cause rkvdec crash.

Change-Id: I3668aa22a3d82af6af2c87ca970028685b8b1960
Signed-off-by: Jung Zhao <jung.zhao@rock-chips.com>
---
 .../video/rockchip/vcodec/vcodec_service.c    | 25 +++++++++++++------
 1 file changed, 17 insertions(+), 8 deletions(-)

diff --git a/drivers/video/rockchip/vcodec/vcodec_service.c b/drivers/video/rockchip/vcodec/vcodec_service.c
index 9236c5e93215..1590b2a164ed 100644
--- a/drivers/video/rockchip/vcodec/vcodec_service.c
+++ b/drivers/video/rockchip/vcodec/vcodec_service.c
@@ -2198,6 +2198,21 @@ static void vcodec_get_reg_freq_default(struct vpu_subdev_data *data,
 					struct vpu_reg *reg)
 {
 	reg->freq = VPU_FREQ_300M;
+
+	if (data->hw_id == HEVC_ID) {
+		if (reg_probe_hevc_y_stride(reg) > 60000)
+			reg->freq = VPU_FREQ_400M;
+	}
+
+	if (reg->type == VPU_PP)
+		reg->freq = VPU_FREQ_400M;
+}
+
+static void vcodec_get_reg_freq_rk3288(struct vpu_subdev_data *data,
+				       struct vpu_reg *reg)
+{
+	vcodec_get_reg_freq_default(data, reg);
+
 	if (reg->type == VPU_DEC || reg->type == VPU_DEC_PP) {
 		if (reg_check_fmt(reg) == VPU_DEC_FMT_H264) {
 			if (reg_probe_width(reg) > 2560) {
@@ -2211,14 +2226,6 @@ static void vcodec_get_reg_freq_default(struct vpu_subdev_data *data,
 			reg->freq = VPU_FREQ_400M;
 		}
 	}
-
-	if (data->hw_id == HEVC_ID) {
-		if (reg_probe_hevc_y_stride(reg) > 60000)
-			reg->freq = VPU_FREQ_400M;
-	}
-
-	if (reg->type == VPU_PP)
-		reg->freq = VPU_FREQ_400M;
 }
 
 static void vcodec_set_freq_default(struct vpu_service_info *pservice,
@@ -2630,6 +2637,8 @@ static void vcodec_set_hw_ops(struct vpu_service_info *pservice)
 		} else if (of_machine_is_compatible("rockchip,rk3126") ||
 				of_machine_is_compatible("rockchip,rk3128")) {
 			pservice->hw_ops->power_on = vcodec_power_on_rk312x;
+		} else if (of_machine_is_compatible("rockchip,rk3288")) {
+			pservice->hw_ops->get_freq = vcodec_get_reg_freq_rk3288;
 		}
 	}
 }
-- 
2.35.3

